File: services/academic-service/src/test/java/com/visor/school/academic/model/TeacherTest.java
package com.visor.school.academic.model;

import org.junit.jupiter.api.Test;
import java.time.LocalDate;
import java.util.List;
import java.util.UUID;
import static org.junit.jupiter.api.Assertions.*;

class TeacherTest {
    @Test
    void shouldCreateTeacherWithRequiredFields() {
        Teacher teacher = new Teacher("EMP-2025-001", UUID.randomUUID(), List.of(), List.of("Mathematics", "Physics"), LocalDate.of(2020, 1, 1), null, EmploymentStatus.ACTIVE);
        assertNull(teacher.getId());
        assertEquals("EMP-2025-001", teacher.getEmployeeId());
        assertEquals(EmploymentStatus.ACTIVE, teacher.getEmploymentStatus());
        assertEquals(2, teacher.getSubjectSpecializations().size());
        assertTrue(teacher.getSubjectSpecializations().contains("Mathematics"));
    }
    @Test
    void shouldRequireAtLeastOneSubjectSpecialization() {
        assertThrows(IllegalArgumentException.class, () -> new Teacher("EMP-2025-002", UUID.randomUUID(), List.of(), List.of(), LocalDate.of(2020, 1, 1), null, EmploymentStatus.ACTIVE));
    }
    @Test
    void shouldAcceptMultipleSubjectSpecializations() {
        Teacher teacher = new Teacher("EMP-2025-003", UUID.randomUUID(), List.of(), List.of("Mathematics", "Physics", "Chemistry"), LocalDate.of(2020, 1, 1), null, EmploymentStatus.ACTIVE);
        assertEquals(3, teacher.getSubjectSpecializations().size());
    }
    @Test
    void shouldAcceptQualifications() {
        Teacher teacher = new Teacher("EMP-2025-004", UUID.randomUUID(), List.of("Bachelor's Degree", "Teaching Certificate"), List.of("English"), LocalDate.of(2020, 1, 1), null, EmploymentStatus.ACTIVE);
        assertEquals(2, teacher.getQualifications().size());
        assertTrue(teacher.getQualifications().contains("Bachelor's Degree"));
    }
    @Test
    void shouldUpdateEmploymentStatus() throws InterruptedException {
        Teacher teacher = new Teacher("EMP-2025-005", UUID.randomUUID(), List.of(), List.of("History"), LocalDate.of(2020, 1, 1), null, EmploymentStatus.ACTIVE);
        assertEquals(EmploymentStatus.ACTIVE, teacher.getEmploymentStatus());
        var initialUpdatedAt = teacher.getUpdatedAt();
        Thread.sleep(10);
        teacher.updateEmploymentStatus(EmploymentStatus.ON_LEAVE);
        assertEquals(EmploymentStatus.ON_LEAVE, teacher.getEmploymentStatus());
        assertTrue(teacher.getUpdatedAt().isAfter(initialUpdatedAt));
    }
    @Test
    void shouldAcceptAllEmploymentStatuses() {
        List<EmploymentStatus> statuses = List.of(EmploymentStatus.ACTIVE, EmploymentStatus.ON_LEAVE, EmploymentStatus.TERMINATED, EmploymentStatus.RETIRED);
        for (EmploymentStatus status : statuses) {
            Teacher teacher = new Teacher("EMP-" + status.name(), UUID.randomUUID(), List.of(), List.of("Test"), LocalDate.of(2020, 1, 1), null, status);
            assertEquals(status, teacher.getEmploymentStatus());
        }
    }
    @Test
    void shouldAcceptDepartmentAssignment() {
        Teacher teacher = new Teacher("EMP-2025-006", UUID.randomUUID(), List.of(), List.of("Science"), LocalDate.of(2020, 1, 1), "Science Department", EmploymentStatus.ACTIVE);
        assertEquals("Science Department", teacher.getDepartment());
    }
}

File: services/academic-service/src/test/kotlin/com/visor/school/academicservice/model/TeacherTest.kt
package com.visor.school.academicservice.model

import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import java.time.LocalDate
import java.util.UUID

class TeacherTest {

    @Test
    fun `should create teacher with required fields`() {
        val teacher = Teacher(
            employeeId = "EMP-2025-001",
            userId = UUID.randomUUID(),
            subjectSpecializations = listOf("Mathematics", "Physics"),
            hireDate = LocalDate.of(2020, 1, 1)
        )

        // ID is null until entity is persisted
        assertNull(teacher.id)
        assertEquals("EMP-2025-001", teacher.employeeId)
        assertEquals(EmploymentStatus.ACTIVE, teacher.employmentStatus)
        assertEquals(2, teacher.subjectSpecializations.size)
        assertTrue(teacher.subjectSpecializations.contains("Mathematics"))
    }

    @Test
    fun `should require at least one subject specialization`() {
        assertThrows<IllegalArgumentException> {
            Teacher(
                employeeId = "EMP-2025-002",
                userId = UUID.randomUUID(),
                subjectSpecializations = emptyList(),
                hireDate = LocalDate.of(2020, 1, 1)
            )
        }
    }

    @Test
    fun `should accept multiple subject specializations`() {
        val teacher = Teacher(
            employeeId = "EMP-2025-003",
            userId = UUID.randomUUID(),
            subjectSpecializations = listOf("Mathematics", "Physics", "Chemistry"),
            hireDate = LocalDate.of(2020, 1, 1)
        )

        assertEquals(3, teacher.subjectSpecializations.size)
    }

    @Test
    fun `should accept qualifications`() {
        val teacher = Teacher(
            employeeId = "EMP-2025-004",
            userId = UUID.randomUUID(),
            qualifications = listOf("Bachelor's Degree", "Teaching Certificate"),
            subjectSpecializations = listOf("English"),
            hireDate = LocalDate.of(2020, 1, 1)
        )

        assertEquals(2, teacher.qualifications.size)
        assertTrue(teacher.qualifications.contains("Bachelor's Degree"))
    }

    @Test
    fun `should update employment status`() {
        val teacher = Teacher(
            employeeId = "EMP-2025-005",
            userId = UUID.randomUUID(),
            subjectSpecializations = listOf("History"),
            hireDate = LocalDate.of(2020, 1, 1),
            employmentStatus = EmploymentStatus.ACTIVE
        )

        assertEquals(EmploymentStatus.ACTIVE, teacher.employmentStatus)
        val initialUpdatedAt = teacher.updatedAt

        Thread.sleep(10)
        teacher.updateEmploymentStatus(EmploymentStatus.ON_LEAVE)

        assertEquals(EmploymentStatus.ON_LEAVE, teacher.employmentStatus)
        assertTrue(teacher.updatedAt.isAfter(initialUpdatedAt))
    }

    @Test
    fun `should accept all employment statuses`() {
        val statuses = listOf(
            EmploymentStatus.ACTIVE,
            EmploymentStatus.ON_LEAVE,
            EmploymentStatus.TERMINATED,
            EmploymentStatus.RETIRED
        )

        statuses.forEach { status ->
            val teacher = Teacher(
                employeeId = "EMP-${status.name}",
                userId = UUID.randomUUID(),
                subjectSpecializations = listOf("Test"),
                hireDate = LocalDate.of(2020, 1, 1),
                employmentStatus = status
            )

            assertEquals(status, teacher.employmentStatus)
        }
    }

    @Test
    fun `should accept department assignment`() {
        val teacher = Teacher(
            employeeId = "EMP-2025-006",
            userId = UUID.randomUUID(),
            subjectSpecializations = listOf("Science"),
            hireDate = LocalDate.of(2020, 1, 1),
            department = "Science Department"
        )

        assertEquals("Science Department", teacher.department)
    }
}

